lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,526 @@
+	notifyTimerResetChan  chan time.Duration
+	notifyChan            chan []string
+	cfg                   *config.Wrapper
+	err                   error
+	errMut                sync.RWMutex
+	ctx                   context.Context
+	cancel                context.CancelFunc
+}
+
+type Service interface {
+	Serve()
+	Stop()
+	C() <-chan []string
+	UpdateIgnores(ignores *ignore.Matcher)
+	VerifyConfiguration(from, to config.Configuration) error
+	CommitConfiguration(from, to config.Configuration) bool
+	String() string
+}
+
+func New(folderCfg config.FolderConfiguration, cfg *config.Wrapper, ignores *ignore.Matcher) Service {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	fsWatcher := &watcher{
+		ignores:               ignores,
+		ignoresUpdate:         make(chan *ignore.Matcher),
+		folderCfgUpdate:       make(chan config.FolderConfiguration),
+		notifyTimerNeedsReset: false,
+		notifyTimerResetChan:  make(chan time.Duration),
+		notifyChan:            make(chan []string),
+		cfg:                   cfg,
+		errMut:                sync.NewRWMutex(),
+		ctx:                   ctx,
+		cancel:                cancel,
+	}
+
+	fsWatcher.updateConfig(folderCfg)
+
+	return fsWatcher
+}
+
+func (w *watcher) Serve() {
+	watchCtx, watchCancel := context.WithCancel(w.ctx)
+	eventChan, err := w.folderCfg.Filesystem().Watch(".", w.ignores, watchCtx, w.folderCfg.IgnorePerms)
+	if err != nil {
+		l.Debugln(w, "failed to setup backend", err)
+		w.errMut.Lock()
+		if err != w.err {
+			l.Warnf("Failed to start filesystem watcher for folder %s: %v", w.folderCfg.Description(), err)
+			w.err = err
+		}
+		w.errMut.Unlock()
+		return
+	}
+
+	w.errMut.Lock()
+	w.err = nil
+	w.errMut.Unlock()
+	l.Infoln("Started filesystem watcher for folder", w.folderCfg.Description())
+
+	// Will not return unless watcher is stopped or an unrecoverable error occurs
+	// Necessary for unit tests where the backend is mocked
+	w.mainLoop(eventChan, watchCancel)
+}
+
+func (w *watcher) mainLoop(eventChan <-chan fs.Event, watchCancel context.CancelFunc) {
+	w.notifyTimer = time.NewTimer(w.notifyDelay)
+	defer w.notifyTimer.Stop()
+
+	inProgress := make(map[string]struct{})
+	inProgressItemSubscription := events.Default.Subscribe(events.ItemStarted | events.ItemFinished)
+
+	w.cfg.Subscribe(w)
+
+	rootEventDir := newEventDir()
+
+	for {
+		select {
+		case event := <-eventChan:
+			w.newEvent(event, rootEventDir, inProgress)
+		case event := <-inProgressItemSubscription.C():
+			updateInProgressSet(event, inProgress)
+		case <-w.notifyTimer.C:
+			w.actOnTimer(rootEventDir)
+		case interval := <-w.notifyTimerResetChan:
+			w.resetNotifyTimer(interval)
+		case ignores := <-w.ignoresUpdate:
+			watchCancel()
+			w.ignores = ignores
+			var err error
+			var watchCtx context.Context
+			watchCtx, watchCancel = context.WithCancel(w.ctx)
+			eventChan, err = w.folderCfg.Filesystem().Watch(".", w.ignores, watchCtx, w.folderCfg.IgnorePerms)
+			if err != nil {
```

Comment:
There is a window between the cancel and the new watch where you will miss events, how is that handled?

Reply:
It isn't handled. All the events buffered in the filesystem are lost. So it is probably a good idea to schedule a full rescan when ignores change.

Reply:
Agree.

Reply:
Also, I wonder if it would not be better to handle this from the model.
Before every startup of watcher, we wire the watcher and follow up by a full scan once it's wired.

Once ignores change, we update the watcher which rewrites (we already do that), and do a full scan after the rewiring.

The only problem would be that UpdateIgnores needs to block until the re-wire returns.

Reply:
I am not sure I got exactly what you want, but I just realised that the ignore patterns aren't needed in the fswatcher package anymore, so I can simplify this. Maybe that resolves this issue.

Reply:
Actually I think I know what you mean (maybe not, but anyway): The whole thing could be even simpler if the fswatcher package doesn't take care of the filesystem watch, but gets the event channel from the model. That would also remove the need to track errors when setting up the filesystem watch. Then however I don't really see the point of running it as a suture service - it could just be controlled by the same context as the filesystem watch.
Does all of this make any sense to you?

Reply:
My point is that if ignores trigger some action in the model (telling the service to update the ignores), it might as well restart the service from scratch moving the complexity of ignore updates to the model. Potentially same with the config updates.
Also every time it restarts/rewires the service it would kick off a full scan post start.

This shifts more responsibilities up to the model, which I guess reduces the indirection, yet I am not sure if this makes it better or worse. Just throwing this out as an idea.

Reply:
See comment in the "main discussion" below for follow up.